File: KnnReal.R
library(caret)
library(class)
library(ggplot2)

source("modelpipe.R")

## source("LoadData.R")
## source("NormalizeData.R")
## source("RestrictData.R")
## source("ExtractYs.R")
load("prepared_datasets.RData")


knnModels = mapply(
    FUN = train,
    xnorms,
    ys,
    MoreArgs = list(
        method = "knn",
        tuneGrid = data.frame(k=3),
        trControl = trainControl(
            method = "cv",
            number = 5,
            seeds = as.list(rep(123, 6))
        )
    ),
    SIMPLIFY = FALSE
)
knnCvAccs = sapply(knnModels, function(u) {u$results$Accuracy})


## -----------------------------------------------------------------
## try with univariate filter feature selection
## -----------------------------------------------------------------
fsKnnFitter = SolderedPipeFitter(
    FastTSelector(nFeat = 10),
    KnnFitter(k = 3)
)

fsKnnFits = mapply(
    FUN = train,
    lapply(1:length(xnorms), function(...) {fsKnnFitter}),
    xnorms,
    ys,
    MoreArgs = list(
        trControl = trainControl(
            method = "cv",
            number = 5,
            seeds = as.list(rep(123, 6))
        )
    ),
    SIMPLIFY = FALSE
)


## -----------------------------------------------------------------
## vary number of features used
## -----------------------------------------------------------------
nFeatures = c(1, 2, 5, 10, 20, 50, 100, 200, 500,
        1000, 2000, 5000, 10000)
names(nFeatures) = as.character(nFeatures)
fitKnnWithNFeat = function(n, setname, fold=5, seed=123) {
    if (n > ncol(xnorms[[setname]])) {
        return(NA)
    }
    fsKnnFitter = SolderedPipeFitter(
        FastTSelector(nFeat = n),
        KnnFitter(k = 3)
    )
    fit = train(
        fsKnnFitter,
        xnorms[[setname]],
        ys[[setname]],
        trControl = trainControl(
            method = "cv",
            number = fold,
            seeds = as.list(rep(seed, times=fold+1))
        )
    )
    return(fit$results$Accuracy)
}

## xnames = names(xnorms)
## names(xnames) = xnames
## accsByNFeats = lapply(
##     X = xnames,
##     FUN = function(s) {
##         lapply(nFeatures, fitKnnWithNFeat, setname=s)
##     }
## )
## accsByNFeats = unlist(accsByNFeats)

## ggdata = data.frame(acc=accsByNFeats, row.names=names(accsByNFeats))
## ggdata$set = factor(gsub("\\..*", "",  names(accsByNFeats)),
##         levels=names(xnorms))
## ggdata$p = as.integer(gsub(".*\\.", "", names(accsByNFeats)))
## write.table(
##     ggdata,
##     file = "KnnRealAccuracyByNFeat_R.tsv",
##     sep = "\t",
##     quote = FALSE,
##     row.names = FALSE
## )

ggdata = read.table(
    "KnnRealAccuracyByNFeat_R.tsv",
    sep = "\t",
    row.names = NULL,
    header = TRUE
)

ggdata$set = factor(as.character(ggdata$set), levels=names(xnorms))
ggobj = ggplot(data=ggdata, mapping=aes(x=p, y=acc, color=set))
ggobj = ggobj + geom_point()
ggobj = ggobj + geom_line(alpha=0.5)
ggobj = ggobj + scale_x_log10(breaks=c(10, 100, 1000, 10000))
ggobj = ggobj + theme_classic()
ggobj = ggobj + scale_color_manual(
        values=c("darkgray", "black", "red", "dodgerblue3"))
ggobj = ggobj + ylab("Accuracy (5-fold CV)")
pdf("KnnRealAccuracyByNFeat.pdf", h=5, w=5*1.325)
print(ggobj)
garbage = dev.off()


## -----------------------------------------------------------------
## use PCA feature extraction
## -----------------------------------------------------------------
feKnnFitter = SolderedPipeFitter(
    PcaExtractor(k = 3),
    KnnFitter(k = 3)
)

## feKnnFits = mapply(
##     FUN = train,
##    lapply(1:length(xnorms), function(...) {feKnnFitter})
##     xnorms,
##     ys,
##     MoreArgs = list(
##         trControl = trainControl(
##             method = "cv",
##             number = 5,
##             seeds = as.list(rep(123, 6))
##         )
##     ),
##     SIMPLIFY = FALSE
## )
## save(feKnnFits, file="feKnnFits.RData")
load("feKnnFits.RData")


## -----------------------------------------------------------------
## test with varying number of principal components
## -----------------------------------------------------------------
npcs = c(1, 2, 5, 10, 20, 50, 100, 200)
names(npcs) = as.character(npcs)
fitKnnWithNPcs = function(n, setname, fold=5, seed=123) {
    if (n > floor((1-1/fold)*min(dim(xnorms[[setname]])))) {
        return(NA)
    }
    feKnnFitter = SolderedPipeFitter(
        PcaExtractor(k = n),
        KnnFitter(k = 3)
    )
    fit = train(
        feKnnFitter,
        xnorms[[setname]],
        ys[[setname]],
        trControl = trainControl(
            method = "cv",
            number = fold,
            seeds = as.list(rep(seed, times=fold+1))
        )
    )
    return(fit$results$Accuracy)
}

## xnames = names(xnorms)
## names(xnames) = xnames
## accsByNPcs = lapply(
##     X = xnames,
##     FUN = function(s) {
##         lapply(npcs, fitKnnWithNPcs, setname=s)
##     }
## )
## accsByNPcs = unlist(accsByNPcs)

## ggdata = data.frame(acc=accsByNPcs, row.names=names(accsByNPcs))
## ggdata$set = factor(gsub("\\..*", "",  names(accsByNPcs)),
##         levels=names(xnorms))
## ggdata$p = as.integer(gsub(".*\\.", "", names(accsByNPcs)))
## write.table(
##     ggdata,
##     file = "KnnRealAccuracyByNPcs_R.tsv",
##     sep = "\t",
##     quote = FALSE,
##     row.names = FALSE
## )

ggdata = read.table(
    "KnnRealAccuracyByNPcs_R.tsv",
    sep = "\t",
    row.names = NULL,
    header = TRUE
)

ggdata$set = factor(as.character(ggdata$set), levels=names(xnorms))
ggobj = ggplot(data=ggdata, mapping=aes(x=p, y=acc, color=set))
ggobj = ggobj + geom_point()
ggobj = ggobj + geom_line(alpha=0.5)
ggobj = ggobj + scale_x_log10(breaks=c(10, 100, 1000, 10000))
ggobj = ggobj + theme_classic()
ggobj = ggobj + scale_color_manual(
        values=c("darkgray", "black", "red", "dodgerblue3"))
ggobj = ggobj + ylab("Accuracy (5-fold CV)")
pdf("KnnRealAccuracyByNPcs.pdf", h=5, w=5*1.325)
print(ggobj)
garbage = dev.off()

File: KnnReal.py
from collections import OrderedDict
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pandas import DataFrame
from pandas import Series
import plotnine as gg
import sklearn as sk
import sklearn.model_selection as model_selection
from sklearn.model_selection import ShuffleSplit
import sklearn.feature_selection as feature_selection
import sklearn.neighbors as neighbors
import sklearn.pipeline as pipeline

import warnings
warnings.filterwarnings("ignore")

import pcaextractor
import MaclearnUtilities
from MaclearnUtilities import safeFactorize

plt.ion()

import RestrictedData
xs = RestrictedData.xs
xnorms = RestrictedData.xnorms
annots = RestrictedData.annots
ys = RestrictedData.ys
ynums = RestrictedData.ynums


def pandaize(f):
    def pandaized(estimator, X, y, **kwargs):
        return f(estimator, np.array(X), safeFactorize(y), **kwargs)
    return pandaized

@pandaize
def cross_val_score_pd(estimator, X, y, **kwargs):
    return model_selection.cross_val_score(
            estimator, X, y, **kwargs)


knnClass = neighbors.KNeighborsClassifier(n_neighbors=3)
cvSchedules = {k : ShuffleSplit(n_splits = 5,
                                test_size = 0.2,
                                random_state = 123)
               for k in xnorms}
knnCvAccs = {k : np.mean(cross_val_score_pd(estimator = knnClass,
                                            X = xnorms[k],
                                            y = ys[k],
                                            cv = cvSchedules[k].split(xnorms[k])))
             for k in xnorms}


## -----------------------------------------------------------------
## try with univariate filter feature selection
## -----------------------------------------------------------------
fsKnnFitter = pipeline.Pipeline([
    ('featsel', feature_selection.SelectKBest(
            feature_selection.f_regression, k=10)),
    ('classifier', neighbors.KNeighborsClassifier(
            n_neighbors=3))
])

fsKnnCvAccs = {k : np.mean(cross_val_score_pd(estimator = fsKnnFitter,
                                              X = xnorms[k],
                                              y = ys[k],
                                              cv = cvSchedules[k].split(xnorms[k])))
               for k in xnorms}


## -----------------------------------------------------------------
## vary number of features used
## -----------------------------------------------------------------
nFeatures = [1, 2, 5, 10, 20, 50, 100, 200, 500,
             1000, 2000, 5000, 10000]
def fitKnnWithNFeat(n, setname, cv=None):
    if cv is None:
        cv = cvSchedules[setname]
    if n > xnorms[setname].shape[1]:
        return None
    fsKnnFitter = pipeline.Pipeline([
        ('featsel', feature_selection.SelectKBest(
                feature_selection.f_regression, k=n)),
        ('classifier', neighbors.KNeighborsClassifier(
                n_neighbors=3))
    ])
    return np.mean(cross_val_score_pd(estimator = fsKnnFitter,
                                      X = xnorms[setname],
                                      y = ys[setname],
                                      cv = cv.split(xnorms[setname])))

accsByNFeats = OrderedDict([(s, OrderedDict([(n, fitKnnWithNFeat(n, s))
                                             for n in nFeatures]))
                            for s in xnorms])

plotData = pd.concat([DataFrame({"set" : s,
                                 "p" : p,
                                 "acc" : accsByNFeats[s][p]},
                                index = [s + "_" + str(p)])
                      for s in accsByNFeats
                      for p in accsByNFeats[s]],
                     axis = 0)
plotData['acc'] = plotData['acc'].astype(float)

plt.close()
ggo = gg.ggplot(plotData, gg.aes(x='p', y='acc', color='set'))
ggo += gg.geom_line()
ggo += gg.scale_x_log10()
ggo += gg.theme_bw()
print(ggo)

# plotData.to_csv("KnnRealAccuracyByNFeat.tsv",
#                 sep = "\t",
#                 index = False,
#                 header = True)


## -----------------------------------------------------------------
## use PCA feature extraction
## -----------------------------------------------------------------
feKnnFitter = pipeline.Pipeline([
    ('featextr', pcaextractor.PcaExtractor(k=3)),
    ('classifier', neighbors.KNeighborsClassifier(
            n_neighbors=3))
])

xmod = feKnnFitter.fit(np.array(xnorms['patel']), np.array(ys['patel']))
xcv = cross_val_score_pd(feKnnFitter, xnorms['patel'], ys['patel'],
                         cv=cvSchedules['patel'].split(xnorms['patel']))

feKnnCvAccs = {k : np.mean(cross_val_score_pd(estimator = feKnnFitter,
                                              X = xnorms[k],
                                              y = ys[k],
                                              cv = cvSchedules[k].split(xnorms[k])))
               for k in xnorms}


## -----------------------------------------------------------------
## test with varying number of principal components
## -----------------------------------------------------------------
npcs = [1, 2, 5, 10, 20, 50, 100, 200]
def fitKnnWithNPcs(n, setname, cv=None):
    if cv is None:
        cv = cvSchedules[setname]
    if n > min(xnorms[setname].shape):
        return None
    feKnnFitter = pipeline.Pipeline([
        ('featextr', pcaextractor.PcaExtractor(k=n)),
        ('classifier', neighbors.KNeighborsClassifier(
                n_neighbors=3))
    ])
    return np.mean(cross_val_score_pd(estimator = feKnnFitter,
                                      X = xnorms[setname],
                                      y = ys[setname],
                                      cv = cv.split(xnorms[setname])))

accsByNPcs = OrderedDict([(s, OrderedDict([(n, fitKnnWithNPcs(n, s))
                                           for n in npcs]))
                            for s in xnorms])

plotData = pd.concat([DataFrame({"set" : s,
                                 "p" : p,
                                 "acc" : accsByNPcs[s][p]},
                                index = [s + "_" + str(p)])
                      for s in accsByNPcs
                      for p in accsByNPcs[s]],
                     axis = 0)
plotData['acc'] = plotData['acc'].astype(float)

plt.close()
ggo = gg.ggplot(plotData, gg.aes(x='p', y='acc', color='set'))
ggo += gg.geom_line()
ggo += gg.scale_x_log10()
ggo += gg.theme_bw()
print(ggo)
